benchmarks/lockhammer/src/lockhammer.c

```diff
@@ -38,73 +38,96 @@
+    fprintf(stderr,
+"Usage: %s\n\t[-t threads]\n\t[-a acquires per thread]\n\t" \
+"[-c critical iterations]\n\t[-p parallelizable iterations]\n\t" \
+"[-- <test specific arguments>]\n", invoc);
+}
+
 int main(int argc, char** argv)
 {
     struct sched_param sparam;
 
     unsigned long i;
-    unsigned long num_cores, num_threads;
-    unsigned long locks_per_thread;
-    unsigned long lock_hold_work, non_lock_work;
+    unsigned long num_cores;
     unsigned long result;
     unsigned long sched_elapsed = 0, real_elapsed = 0;
     unsigned long start_ns = 0;
     double avg_lock_depth = 0.0;
 
     num_cores = sysconf(_SC_NPROCESSORS_ONLN);
 
-    if (argc == 1) {
-        num_threads = num_cores;
-        locks_per_thread = 50000;
-        lock_hold_work = 0;
-        non_lock_work = 0;
-    }
-    else if (argc == 5) {
-        num_threads = atoi(argv[1]);
-        /* Do not allow number of threads to exceed online cores
-           in order to prevent deadlock ... */
-        num_threads = num_threads > num_cores ? num_cores : num_threads;
-        locks_per_thread = atoi(argv[2]);
-        lock_hold_work = atoi(argv[3]);
-        non_lock_work = atoi(argv[4]);
-    }
-    else {
-        fprintf(stderr, "Usage: %s [<cores> <threads per core> <critical loops> <post-release loops>]\n", argv[0]);
-        return 1;
+    /* Set defaults for all command line options */
+    test_args args = { .nthrds = num_cores,
+                       .nacqrs = 50000,
+                       .ncrit = 0,
+                       .nparallel = 0 };
+
+    for (i = 1; i < argc; i++) {
+        if (strlen(argv[i]) == 2 && argv[i][0] == '-' && i < argc) {
+            switch (argv[i][1]) {
+              case 't':
+                i++;
+                args.nthrds = strtol(argv[i], (char **) NULL, 10);
+                /* Do not allow number of threads to exceed online cores
+                   in order to prevent deadlock ... */
+                args.nthrds > num_cores ? num_cores : args.nthrds;
+                break;
+              case 'a':
+                i++;
+                args.nacqrs = strtol(argv[i], (char **) NULL, 10);
+                break;
+              case 'c':
+                i++;
+                args.ncrit = strtol(argv[i], (char **) NULL, 10);
+                break;
+              case 'p':
+                i++;
+                args.nparallel = strtol(argv[i], (char **) NULL, 10);
+                break;
+              case '-':
+                /* anything after "--" is ignore and passed through to test */
+                i++;
+                parse_test_args(args, argc - i, &argv[i]);
+                /* skip over remaining args since they are parsed by test */
+                i = argc;
+                break;
+              default:
+                print_usage(argv[0]);
+                return 1;
+            }
+            if (errno == EINVAL) {
+                print_usage(argv[0]);
+                return 1;
+            }
```

Comment:
Instead of writing our own command line parser, why not use getopt?  When getopt hits a "--" it will stop parsing and we can then parse the remainder if needed in your optional parse_test_args function per test.